lib/block_view_lockups.go

```diff
@@ -0,0 +1,2178 @@
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupInvalidTransactorPKID, "_connectCoinLockup")
+	}
+	hodlerPKID := transactorPKIDEntry.PKID
+
+	// Determine which profile PKID to use.
+	var profilePKID *PKID
+	if txMeta.ProfilePublicKey.IsZeroPublicKey() {
+		profilePKID = ZeroPKID.NewPKID()
+	} else {
+		profilePKIDEntry := bav.GetPKIDForPublicKey(txMeta.ProfilePublicKey.ToBytes())
+		if profilePKIDEntry == nil || profilePKIDEntry.isDeleted {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupNonExistentProfile, "_connectCoinLockup")
+		}
+		profilePKID = profilePKIDEntry.PKID.NewPKID()
+	}
+
+	// Validate the transactor as having sufficient DAO Coin or DESO balance for the transaction.
+	var transactorBalanceNanos256 *uint256.Int
+	var prevTransactorBalanceEntry *BalanceEntry
+	var prevCoinEntry *CoinEntry
+	if profilePKID.IsZeroPKID() {
+		// NOTE: When spending balances, we need to check for immature block rewards. Since we don't have
+		//       the block rewards yet for the current block, we subtract one from the current block height
+		//        when spending balances.
+
+		// Check the DeSo balance of the user.
+		transactorBalanceNanos, err := bav.GetSpendableDeSoBalanceNanosForPublicKey(txn.PublicKey, blockHeight-1)
+		if err != nil {
+			return 0, 0, nil, errors.Wrap(err, "_connectCoinLockup")
+		}
+
+		// Construct a uint256 balance and validate the transactor as having sufficient DeSo.
+		transactorBalanceNanos256, _ = uint256.FromBig(big.NewInt(0).SetUint64(transactorBalanceNanos))
+		if txMeta.LockupAmountBaseUnits.Gt(transactorBalanceNanos256) {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupInsufficientDeSo, "_connectCoinLockup")
+		}
+
+		// Spend the transactor's DeSo balance.
+		lockupAmount64 := txMeta.LockupAmountBaseUnits.Uint64()
+		newUtxoOp, err := bav._spendBalance(lockupAmount64, txn.PublicKey, blockHeight-1)
+		if err != nil {
+			return 0, 0, nil, errors.Wrapf(err, "_connectCoinLockup")
+		}
+		utxoOpsForTxn = append(utxoOpsForTxn, newUtxoOp)
+	} else {
+		// Check the BalanceEntry of the user.
+		transactorBalanceEntry, _, _ := bav.GetBalanceEntryForHODLerPubKeyAndCreatorPubKey(
+			txn.PublicKey,
+			txMeta.ProfilePublicKey.ToBytes(),
+			true)
+		if transactorBalanceEntry == nil || transactorBalanceEntry.isDeleted {
+			return 0, 0, nil,
+				errors.Wrapf(RuleErrorCoinLockupBalanceEntryDoesNotExist, "_connectCoinLockup")
+		}
+
+		// Validate the balance entry as having sufficient funds.
+		transactorBalanceNanos256 = transactorBalanceEntry.BalanceNanos.Clone()
+		if txMeta.LockupAmountBaseUnits.Gt(transactorBalanceNanos256) {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupInsufficientCoins, "_connectCoinLockup")
+		}
+
+		// We store the previous transactor balance entry in the event we need to revert the transaction.
+		prevTransactorBalanceEntry = transactorBalanceEntry.Copy()
+
+		// Spend the transactor's DAO coin balance.
+		transactorBalanceEntry.BalanceNanos =
```

Comment:
I think it's a good idea to Copy() it before you modify it and set it. It just prevents weird stuff from happening in general.

Reply:
Eh you copied the Prev one so it's probably fine.